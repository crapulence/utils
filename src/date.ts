/**
 * 获取给定年月的最后一天
 * @param year number
 * @param month number
 * @returns number
 */
export function getMonthEndDay(year: number, month: number): number {
  return 32 - new Date(year, month - 1, 32).getDate()
}
