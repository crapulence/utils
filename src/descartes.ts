export function descartes<T>(groups: T[][]): T[][] {
  return groups.reduce<T[][]>((a, b) => a.flatMap(x => b.map(y => [...x, y])), [[]])
}
