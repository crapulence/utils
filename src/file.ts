export function getFileExtension(filename: string): string {
  return filename.slice((filename.lastIndexOf('.') - 1 >>> 0) + 2)
}

export type ResultType = 'dataUrl' | 'text' | 'file'
export function readFile(file: File, resultType: ResultType): Promise<string | null> {
  return new Promise((resolve, reject) => {
    if (resultType === 'file') {
      resolve(null)
      return
    }

    const reader = new FileReader()

    reader.onload = (event) => {
      const result = event.target?.result as string | null
      resolve(result)
    }

    reader.onerror = (error) => {
      reject(error)
    }

    if (resultType === 'dataUrl') {
      reader.readAsDataURL(file)
    }
    else if (resultType === 'text') {
      reader.readAsText(file)
    }
    else {
      reject(new Error(`Invalid result type: ${resultType}`))
    }
  })
}
