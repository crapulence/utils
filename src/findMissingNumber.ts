/**
 * 找出连续数字数组中缺失的数字
 * @param {Array} nums
 */
export function findMissingNumber(nums: number[]) {
  return Array.from(Array(Math.max(...nums)).keys())
    .map((_, i) => (!nums.includes(i) ? i : null))
    .filter(f => f)
}
