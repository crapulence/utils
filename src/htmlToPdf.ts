import html2canvas from 'html2canvas'
import { jsPDF } from 'jspdf'

export function htmlToPdf(element: HTMLElement): Promise<jsPDF> {
  return new Promise((resolve) => {
    html2canvas(element, {
      y: 300,
    }).then((canvas) => {
      const contentWidth = canvas.width
      const contentHeight = canvas.height
      const pageHeight = (contentWidth / 592.28) * 841.89
      let leftHeight = contentHeight - 100
      let position = 0
      const imgWidth = 595.28
      const imgHeight = (592.28 / contentWidth) * contentHeight
      const pageData = canvas.toDataURL('image/jpeg', 1.0)
      // eslint-disable-next-line new-cap
      const pdf = new jsPDF({
        unit: 'pt',
      })

      if (leftHeight < pageHeight) {
        pdf.addImage(pageData, 'JPEG', 0, 0, imgWidth, imgHeight)
      }
      else {
        while (leftHeight > 0) {
          pdf.addImage(pageData, 'JPEG', 0, position, imgWidth, imgHeight)
          leftHeight -= pageHeight
          position -= 841.89

          if (leftHeight > 0) {
            pdf.addPage()
          }
        }
      }

      resolve(pdf)
    })
  })
}
