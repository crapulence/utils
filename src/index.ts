export * from './date'
export * from './descartes'
export * from './file'
export * from './htmlToPdf'
export * from './number'
export * from './swap'
export * from './system'
export * from './tree'
