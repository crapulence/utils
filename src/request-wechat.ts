import qs from 'qs'
import { SERVER_URL } from './constants'

type Options = Omit<WechatMiniprogram.RequestOption, 'success' | 'fail'>

interface Response<T> {
  code: number
  data: T
  msg: string
}

function createRequest<T>(options: Options) {
  return new Promise<Response<T>>((resolve, reject) => {
    wx.request<Response<T>>({
      ...options,
      url: `${SERVER_URL}${options.url}`,
      header: {
        Authorization: wx.getStorageSync('token'),
      },
      success: (res) => {
        if (res.data.code === 0) {
          resolve(res.data)
        }
        else {
          if (res.data.code === 401) {
            wx.reLaunch({
              url: '/pages/login/index',
              fail: (err) => {
                // eslint-disable-next-line no-console
                console.log(err)
              },
            })
          }
          reject(res.data.msg)
        }
      },
      fail: (err) => {
        reject(err.errMsg)
      },
    })
  })
}

function get<T>(
  url: string,
  data?: Record<string, any>,
  options?: Partial<Options>,
) {
  return createRequest<T>({
    method: 'GET',
    url: `${url}?${qs.stringify(data, { arrayFormat: 'repeat' })}`,
    ...options,
  })
}

function post<T>(
  url: string,
  data?: Record<string, any>,
  options?: Partial<Options>,
) {
  return createRequest<T>({
    method: 'POST',
    url,
    data,
    ...options,
  })
}

function put<T>(
  url: string,
  data?: Record<string, any>,
  options?: Partial<Options>,
) {
  return createRequest<T>({
    method: 'PUT',
    url,
    data,
    ...options,
  })
}

function del<T>(
  url: string,
  data?: Record<string, any>,
  options?: Partial<Options>,
) {
  return createRequest<T>({
    method: 'DELETE',
    url: `${url}?${qs.stringify(data, { arrayFormat: 'repeat' })}`,
    ...options,
  })
}

export const request = {
  get,
  post,
  put,
  del,
}
