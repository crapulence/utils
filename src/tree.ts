/**
 * 获取树形结构中满足条件的路径数组
 * @param tree 树形数组
 * @param fn 条件函数
 * @returns 满足条件的路径数组
 */
export function getPathByTree<T extends { children?: T[] }>(
  tree: T[],
  fn: (data: T) => boolean,
  path: T[] = [],
): T[] {
  if (!tree || tree.length === 0) {
    return []
  }

  for (const data of tree) {
    path.push(data)
    if (fn(data)) {
      return path
    }
    if (data.children) {
      const findChildren = getPathByTree(data.children, fn, path)
      if (findChildren.length > 0) {
        return findChildren
      }
    }
    path.pop()
  }
  return []
}

/**
 * 获取多个树形结构中满足条件的路径数组
 * @param trees 树形数组
 * @param fn 条件函数
 * @returns 满足条件的路径数组
 */
export function getPathByTrees<T extends { children?: T[] }>(
  trees: T[],
  fn: (data: T) => boolean,
  path: T[] = [],
  results: T[][] = [],
): T[][] {
  if (!trees || trees.length === 0) {
    return results
  }

  for (const tree of trees) {
    path.push(tree)
    if (fn(tree)) {
      results.push([...path])
    }
    if (tree.children) {
      getPathByTrees(tree.children, fn, path, results)
    }
    path.pop()
  }
  return results
}

/**
 * 遍历树节点
 */
export function walk<T extends { children?: T[] }>(list: T[], cb: (child: T, parent?: T) => unknown, parent?: T): void {
  for (const item of list) {
    cb(item, parent)
    if (item.children?.length) {
      walk(item.children, cb, item)
    }
  }
}
