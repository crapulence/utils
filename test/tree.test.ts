import { expect, it } from 'vitest'
import { getPathByTree, getPathByTrees } from '../src/tree'

const data = [
  {
    label: 'a',
    value: '1',
    children: [
      {
        label: 'a-1',
        value: '1-1',
        children: [],
      },
      {
        label: 'a-2',
        value: '1-2',
        children: [
          {
            label: 'a-2-1',
            value: '1-2-1',
            children: [],
          },
        ],
      },
      {
        label: 'a-3',
        value: '1-3',
        children: [
          {
            label: 'a-3-1',
            value: '1-3-1',
            children: [],
          },
        ],
      },
    ],
  },
]

const data2 = [{ label: 'a', value: 1, children: [{ label: 'a-1', value: 2, children: [] }] }, { label: 'a', value: 1, children: [{ label: 'a-2', value: 2, children: [] }] }]

it('should return the correct array of paths', () => {
  expect(getPathByTree(data, item => item.label === 'a-2-1')).toMatchSnapshot()
})

it('should return multiple correct path arrays', () => {
  expect(getPathByTrees(data2, item => item.label === 'a')).toMatchSnapshot()
})
