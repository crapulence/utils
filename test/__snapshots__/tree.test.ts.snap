// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should return multiple correct path arrays 1`] = `
[
  [
    {
      "children": [
        {
          "children": [],
          "label": "a-1",
          "value": 2,
        },
      ],
      "label": "a",
      "value": 1,
    },
  ],
  [
    {
      "children": [
        {
          "children": [],
          "label": "a-2",
          "value": 2,
        },
      ],
      "label": "a",
      "value": 1,
    },
  ],
]
`;

exports[`should return the correct array of paths 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "label": "a-1",
        "value": "1-1",
      },
      {
        "children": [
          {
            "children": [],
            "label": "a-2-1",
            "value": "1-2-1",
          },
        ],
        "label": "a-2",
        "value": "1-2",
      },
      {
        "children": [
          {
            "children": [],
            "label": "a-3-1",
            "value": "1-3-1",
          },
        ],
        "label": "a-3",
        "value": "1-3",
      },
    ],
    "label": "a",
    "value": "1",
  },
  {
    "children": [
      {
        "children": [],
        "label": "a-2-1",
        "value": "1-2-1",
      },
    ],
    "label": "a-2",
    "value": "1-2",
  },
  {
    "children": [],
    "label": "a-2-1",
    "value": "1-2-1",
  },
]
`;
